src/gleam/uri.gleam

```diff
@@ -164,3 +168,133 @@ pub fn to_string(uri: Uri) -> String {
   |> dynamic.string
   |> result.unwrap("")
 }
+
+/// Fetch the origin of a uri
+///
+/// Return the origin of a uri as defined in
+/// https://tools.ietf.org/html/rfc6454
+///
+/// The supported uri schemes are `http` and `https`
+/// Urls without a scheme will return Error
+pub fn origin(uri: Uri) -> Result(String, Nil) {
+  let Uri(scheme: scheme, host: host, port: port, ..) = uri
+  case scheme {
+    Some("https") | Some("http") -> {
+      let origin = Uri(scheme, None, host, port, "", None, None)
+      Ok(to_string(origin))
+    }
+    _ -> Error(Nil)
+  }
+}
+
+fn option_or(first: Option(a), second: Option(a)) -> Option(a) {
```

Comment:
This could be a function on the option module? 
https://doc.rust-lang.org/std/option/enum.Option.html#method.or

Reply:
Sounds good!